contracts/src/L1/rollup/ScrollChain.sol

```diff
@@ -311,68 +280,57 @@ contract ScrollChain is OwnableUpgradeable, PausableUpgradeable, IScrollChain {
 
-        // verify batch
-        IRollupVerifier(verifier).verifyAggregateProof(_batchIndex, _aggrProof, _publicInputHash);
+        // finalize batch
+        _finalizeBatch(memPtr, _batchHash, _postStateRoot, _withdrawRoot, _aggrProof);
+    }
 
-        // check and update lastFinalizedBatchIndex
-        unchecked {
-            require(lastFinalizedBatchIndex + 1 == _batchIndex, "incorrect batch index");
-            lastFinalizedBatchIndex = _batchIndex;
+    /// @inheritdoc IScrollChain
+    ///
+    /// @dev This function can by used to commit and finalize a new batch in a
+    /// single step if all previous batches are finalized. It can also be used
+    /// to finalize the earliest pending batch. In this case, the provided batch
+    /// should match the pending batch.
+    ///
+    /// If user choose to finalize a pending batch, the batch hash of current
+    /// header should match with `committedBatches[currentIndex]`.
+    /// Otherwise, `committedBatches[currentIndex]` should be `bytes32(0)`.
+    function commitAndFinalizeBatchEnforced(
+        bytes calldata _parentBatchHeader,
+        bytes[] memory _chunks,
+        bytes calldata _skippedL1MessageBitmap,
+        bytes32 _postStateRoot,
+        bytes32 _withdrawRoot,
+        bytes calldata _withdrawRootProof,
+        bytes calldata _aggrProof
+    ) external {
+        // check and enable enforced mode.
+        if (finalizationState.mode == 0) {
+            if (finalizationState.timestamp + maxFinalizationDelay < block.timestamp) {
+                finalizationState.mode = 1;
+            } else {
+                revert("not allowed");
+            }
         }
 
-        // record state root and withdraw root
-        finalizedStateRoots[_batchIndex] = _postStateRoot;
-        withdrawRoots[_batchIndex] = _withdrawRoot;
-
-        // Pop finalized and non-skipped message from L1MessageQueue.
-        uint256 _l1MessagePopped = BatchHeaderV0Codec.l1MessagePopped(memPtr);
-        if (_l1MessagePopped > 0) {
-            IL1MessageQueue _queue = IL1MessageQueue(messageQueue);
+        (uint256 memPtr, bytes32 _batchHash) = _commitBatch(_parentBatchHeader, _chunks, _skippedL1MessageBitmap);
 
-            unchecked {
-                uint256 _startIndex = BatchHeaderV0Codec.totalL1MessagePopped(memPtr) - _l1MessagePopped;
-
-                for (uint256 i = 0; i < _l1MessagePopped; i += 256) {
-                    uint256 _count = 256;
-                    if (_l1MessagePopped - i < _count) {
-                        _count = _l1MessagePopped - i;
-                    }
-                    uint256 _skippedBitmap = BatchHeaderV0Codec.skippedBitmap(memPtr, i / 256);
-
-                    _queue.popCrossDomainMessage(_startIndex, _count, _skippedBitmap);
-
-                    _startIndex += 256;
-                }
-            }
-        }
+        (bytes32 stateRoot, bytes32 storageValue) = IZkTrieVerifier(zkTrieVerifier).verifyZkTrieProof(
```

Comment:
Is it possible that someone submits a valid proof with an invalid withdraw root?